### ISSUE TITLE
Suggestion: Restyle and implement inherited "tab groups" feature 

### ISSUE THREAD

**Issue by Brave-Matt** [brave/brave-browser#7142]
## Description 

It seems as though we've inherited Chrome's experimental "tab groups" feature and that it works in Brave as well without issue (seemingly). The feature still sits be hind a flag (`#tab-groups`) right now and I think it would be advantageous to hijack this feature and make it our own.

![image](https://user-images.githubusercontent.com/3299098/69768355-1adf4480-1135-11ea-861b-3e3b74ca20ea.png)

### Functionality
The way the feature is currently implemented:

**Right-clicking on a tab** gives you options to `Add`, `Add to existing` or `Remove` a tab to/from a group. 

![groups](https://user-images.githubusercontent.com/3299098/69767457-334d6000-1131-11ea-8a0d-14e339d0e942.jpeg)

**You can also drag tabs** into (or out of) pre-existing groups.

![tabdrag](https://user-images.githubusercontent.com/3299098/69767667-3432c180-1132-11ea-9d0a-9da4eb7101fe.gif)

**If you `Ctrl/Cmd + Click` a link** on a tab already grouped, the link will open and be automatically added to that same tab group.

![tabcmdclick](https://user-images.githubusercontent.com/3299098/69767870-16199100-1133-11ea-89da-2fcb1c56a2b9.gif)

**Clicking the tab group color** gives you options for customizing that groups color and giving each group a name/title.

![image](https://user-images.githubusercontent.com/3299098/69767968-7c061880-1133-11ea-9ca1-34740463f9c1.png)

Optionally/additionally, we may want to consider re-implementing the tab "pages" feature in conjunction with this, the way we did in Muon. I'm personally a fan of [Cliftons "slick animation" example](https://github.com/brave/brave-browser/issues/35#issuecomment-414196827). 

## Miscellaneous Information:

Implementing this would make many users happy -- as it has been requested several times and in many different ways. These are just the ones I've pulled from Community in my initial search:

- https://community.brave.com/t/restore-original-tab-grouper/44001
- https://community.brave.com/t/tab-group-feature/45134
- https://community.brave.com/t/auto-tab-grouping-on-tab-bar/21811
- https://community.brave.com/t/brave-chromium-tab-handling/39627
- https://community.brave.com/t/tab-groups-extension-request/6273
- https://community.brave.com/t/group-tabs-on-top-of-tabs-ui/70603
- https://community.brave.com/t/showing-4-tabs-within-one-tab/18132

**Comment by srirambv** [comment]
Finally, my suggestion from [browser-laptop](https://github.com/brave/browser-laptop/issues/9976) is implemented upstream. I'd give a +1 for tab colours so helpful when you have loads of tabs opened

**Comment by bsclifton** [comment]
+1 from @jaredready via https://github.com/brave/brave-browser/issues/9419

**Comment by bsclifton** [comment]
https://www.cnet.com/how-to/google-chrome-now-groups-browser-tabs-heres-how-to-do-it

**Comment by PierreMardon** [comment]
Quick testimony : I almost hesitate to switch back to Chrome to benefit from this feature that I think would bring relief to any many-tabs-worker on a daily basis (don't know if I'm the only one that eager to have it available).

**Comment by Brave-Matt** [comment]
+1 from reddit:
https://www.reddit.com/r/brave_browser/comments/isq30w/tab_groups_aesthetically_different_from_google/

**Comment by jtbrown3** [comment]
The new 'Tab Groups' feature works really great in the Google Chrome browser - and really falls flat in Brave.  Really hoping this feature will be properly implemented in Brave _**soon**_!

**Comment by Sojiro84** [comment]
After trying out the Chromium's tab group feature with the sexy circular design and proper spacing vs Brave's square design with improper spacing (especially where the name of the tab-group color collides with the tab itself), I really prefer Chromium's version.

I hope you guys will fix the spacing. Right now it's visually unpleasant and since I use tab groups a lot, I am back to using Chromium until this is fixed.

**Comment by timelytree** [comment]
Tab groups have proven to be a very powerful and useful feature. Although functionally, they work well in Brave - visually as mentioned above, they're quite unpleasant and I do find myself missing the zen beauty of Chrome's UI implementation for them.

---

Actually this seems like it would be a pretty easy fix if we ever get something like a `custom.css` file as mentioned in https://github.com/brave/browser-laptop/issues/1122 (moved to #774).

> Customization similar to Atom; a simple doc you can edit

---

Adding some screenshots for reference:

**Brave:**
<img width="523" alt="Screen Shot 2021-02-22 at 3 48 54 AM" src="https://user-images.githubusercontent.com/11708190/108685669-6d3b1c80-74c2-11eb-9024-4fad1c4cfb12.png">

**Chrome:**
<img width="463" alt="Screen Shot 2021-02-22 at 3 49 42 AM" src="https://user-images.githubusercontent.com/11708190/108685681-70cea380-74c2-11eb-8f44-0e194187fb03.png">

**Comment by urbenlegend** [comment]
I agree with @timelytree. The functionality works, but the square group name background and the extra line of pixels underneath each tab make it an eyesore.

**Comment by vis2x** [comment]
I also agree with @timelytree. Is there any ETA on when this will be added to Brave?

**Comment by urbenlegend** [comment]
Any progress on this? Seems like it'd be a simple fix and would go a long way to making this feature actually usable in Brave.

**Comment by tunix** [comment]
I look forward for a fix to this issue. It looks really ugly in Brave 😞

**Comment by CumpsD** [comment]
If the text is a little long, like `Financial`, you hit the borders and the left/right padding is different from top/bottom, making it very ugly :(

**Comment by sudokai** [comment]
I'm a heavy tab groups user and, as petty as it sounds, I currently don't use Brave as my main browser because of this 😂

**Comment by jdsimcoe** [comment]
I'm with you, the fact that this hasn't been updated in years is keeping from using Brave as my primary browser.

**Comment by sangwoo108** [comment]
cc @aguscruiz and @zenparsing who recently updated tab style.

**Comment by aguscruiz** [comment]
We have a new style for groups that's coming with the new tab shape design that addresses this 🌈

If anyone wants to take it for a spin before it's official, the flag is `#brave-horizontal-tabs-update`

<img width="528" alt="image" src="https://github.com/brave/brave-browser/assets/15734630/ee1ee493-0ec2-4f25-a89c-96cf22782638">

**Comment by OceanCity** [comment]
I can't believe I just recently, in January of 2025, discovered 'Tab Groups' in Brave!
I LOVE this. Have wanted this kind of feature now for years. And there it is, wow.

So, after poking around, I get the idea, and it's stable, well thought out, however I do have several 'wish list' requests for the developers to make tab groups even better.

1) Please don't add the tab groups to the bookmarks bar. Why? They are _already_ in the top line of open tabs, and adding them to the bookmarks bar impinges on the bookmarks. Sorry, don't mean to nit picky.

2) Unless one is consolidating more than say 2-3-4 tabs, the one new resulting tab group tab takes up almost as much space. To this end I suggest removing the circular icon, and better yet, allow the user to pick or generate a much shorter tab of either a graphic or perhaps just use colors (which of course are already usable now).  

3) Allow the expanded tabs to be 'edited' as are normal tabs. For example, because I use so many open tabs I almost always eliminate some or all of the name or the description. This does not appear possible to edit a tab that's in a tab group.

But, overall, I say KUDOS to tab groups, and I love them, even as they are somewhat gangly at the moment and could be tightened up. Please don't take my comments the wrong way. I am thrilled to have stumbled upon this feature. I just like to maximize my customizability to fit as much as possible on the top line.

Keep up the great work!

All the best,
John

**Comment by rebron** [comment]
Going to close this one as stale as we've Tab Groups in Brave for awhile now. 

> I can't believe I just recently, in January of 2025, discovered 'Tab Groups' in Brave! I LOVE this. Have wanted this kind of feature now for years. And there it is, wow.
> 
> So, after poking around, I get the idea, and it's stable, well thought out, however I do have several 'wish list' requests for the developers to make tab groups even better.
> 
> 1. Please don't add the tab groups to the bookmarks bar. Why? They are _already_ in the top line of open tabs, and adding them to the bookmarks bar impinges on the bookmarks. Sorry, don't mean to nit picky.

There's an option to show or hide Tab groups in the bookmarks bar. brave://settings/appearance

> 3. Unless one is consolidating more than say 2-3-4 tabs, the one new resulting tab group tab takes up almost as much space. To this end I suggest removing the circular icon, and better yet, allow the user to pick or generate a much shorter tab of either a graphic or perhaps just use colors (which of course are already usable now).

You can have just a color to indicate the tab group without adding a label. 

<img width="135" height="50" alt="Image" src="https://github.com/user-attachments/assets/43a31a79-6405-48d3-8b5d-753461b997bf" />

> 5. Allow the expanded tabs to be 'edited' as are normal tabs. For example, because I use so many open tabs I almost always eliminate some or all of the name or the description. This does not appear possible to edit a tab that's in a tab group.> > But, overall, I say KUDOS to tab groups, and I love them, even as they are somewhat gangly at the moment and could be tightened up. Please don't take my comments the wrong way. I am thrilled to have stumbled upon this feature. I just like to maximize my customizability to fit as much as possible on the top line.
> 
We have a flag in Nightly or version 1.83.x now 
brave://flags/#brave-renaming-tabs We still have some work to do with it we'll have that feature as well. 

> Keep up the great work!
> 
> All the best, John
